### PR TITLE
initial refactor for TypeScript

### PR DESCRIPTION
**Pull request by chazkiker2**
# Description

This PR presents an initial refactor for TypeScript support

Includes fixes for the following issues: #3 & #4 

## Featured Changes 

- [x] add initial refactor for a TS implementation of `RyderSerial`
- [x] add TS configs and dependencies
- [x] add build script to generate JS file in `out/index.js` 
- [x] change `outDir` to `dist` and untrack.
- [x] resolve `false` vs `Promise.resolve` in `lock()` and `unlock()` methods
- [x] include type definitions in build (via auto-generated `.d.ts` files)
- [x] resolve inconsistent casing for `Options` interface.
  - I switched the `Options` fields to snake_case. This may justify a `README.md` update and may introduce a breaking change. But at least our casings are all matching now!
- [x] add more scalable support for `logging` as well as more explicit (and more consistent) debug messages. (My hope is that anyone who turns on debug will have a much easier time looking through the code and understanding why _X_ is happening if they do _Y_)
- [x] refactor file structure to support the growing library 🤗 
- [x] configure `eslint` and `prettier`

## future improvements / community conversations & questions

- Leaving `TrainEntry[4]` as type `string` for now as refactoring will introduce a breaking change. More importantly, this refactor is not time-sensitive. The train will (hopefully) undergo some more robust refactors in the future; no need to optimize too early. We should circle back to this conversation when we approach the `train` / sequencer implementation

### CODE REVIEW

**Comment by chazkiker2** [comment]
@MarvinJanssen would love your thoughts on this first attempt. Definitely expecting changes to be required. 

what works? what doesn't? 

let me know! :)

**Comment by chazkiker2** [comment]
# Update

Tested NPM installation with CLI, and I believe everything is working as it was previously!! 

Test file is just a manual test for the time being; didn't know what framework we wanted to use so I didn't add one in yet.

File: out/index.js [hunk 1]
```diff
@@ -0,0 +1,355 @@
+'use strict';
```
Comment (chazkiker2) [inline out/index.js:1]: this is automatically built from typescript, no need to review this file in-depth.

File: src/index.ts [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+      this.next();
+    });
+  };
+
+  next(): void {
+    if (this[state_symbol] === STATE_IDLE) {
+      if (!this[train_symbol].length)
+        return;
+      if (!this.serial || !this.serial.isOpen) {
+        const [, , reject] = this[train_symbol][0];
```
Comment (chazkiker2) [inline src/index.ts:338]: fix for issue #4 found here

File: src/index.ts [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+      var [, , reject] = this[train_symbol].shift()!;
+      reject(error);
+    }
+    clearTimeout(this[watchdog_symbol]);
+    this[state_symbol] = STATE_IDLE;
+    this.next();
+  }
+
+  serial_data(data: Uint8Array): boolean | void {
+    this.options.debug && console.debug('data from Ryder', '0x' + data.toString());
```
Comment (chazkiker2) [inline src/index.ts:124]: this used to feature `data.toString('hex')` which TypeScript ruled out. Should the `data` param be something other than type `Uint8Array`?

Comment (chazkiker2) [reply src/index.ts]: fixed in 86fc02f.

**Comment by chazkiker2** [comment]
@MarvinJanssen this is ready for review & merge, in case that wasn't clear. No rush.

File: src/index.ts [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+  };
+
+  locked(): boolean {
+    return !!this[lock_symbol].length;
+  };
+
+  lock(): Promise<void> {
+    this.options.debug && console.debug('ryderserial lock');
+    if (!this[lock_symbol].length) {
+      // TODO: fix this[lock_symbol]... why is it holding 'false'?
```
Comment (MarvinJanssen) [inline src/index.ts:295]: Just to not construct a pointless Promise. Then `unlock()` can be refactored too.

File: src/index.ts [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+    });
+  };
+
+  next(): void {
+    if (this[state_symbol] === STATE_IDLE) {
+      if (!this[train_symbol].length)
+        return;
+      if (!this.serial || !this.serial.isOpen) {
+        const [, , reject] = this[train_symbol][0];
+        this.clear();
```
Comment (MarvinJanssen) [inline src/index.ts:340]: I wonder if we can to keep the queued commands for when the device reconnects or not. I think this is fine for now and we can decide based on future use-cases. I would imagine that an app dev would show a warning if the device becomes disconnected and for the user to restart whatever process it was in.

File: test/index.js [hunk 6]
```diff
@@ -0,0 +1,43 @@
+"use strict";
+
+const RyderSerial = require("../out/index");
```
Comment (MarvinJanssen) [inline test/index.js:3]: I think we should have a test file at some point, however I am not a huge fan of this one because it cannot be automated / has hardcoded values. We probably need a way to deliver the simulator and run it in headless mode before unit tests make sense. What do you think?

File: package.json [hunk 7]
```diff
@@ -2,21 +2,32 @@
   "name": "ryderserial-proto",
   "version": "0.0.1",
   "description": "",
-  "main": "src/index.js",
+  "main": "out/index.js",
```
Comment (MarvinJanssen) [inline package.json:5]: I think we should make it build to a `dist` folder including the typedefs and untrack the folder. That would be great for downstream implementations.

**Review by MarvinJanssen — CHANGES_REQUESTED** [review]
Looks great so far, added a few comments.

File: src/index.ts [hunk 8]
```diff
@@ -0,0 +1,384 @@
...
+const train_symbol = Symbol('train');
+const watchdog_symbol = Symbol('watchdog');
+const reconnect_symbol = Symbol('reconnect');
+
+const WATCHDOG_TIMEOUT = 5000;
+
+var id = 0;
+
+export interface RyderSerialOptions extends SerialPort.OpenOptions {
+  rejectOnLocked: any;
```
[the remark this answers is not in the document]
Comment (chazkiker2) [reply src/index.ts:48]: totally a fan of consistency. And I prefer snake in general, but I usually follow camelCase in my own code when it's TS/JS. I'd imagine most contributors to this package would likewise follow JS conventions.

Left this one as camel b/c it was camel previously (and noted that way in all the `README` examples). Didn't want to overstep on what was here... 

I will say that changing these to snake case will break consistency as far as properties on the `RyderSerialOptions` is concerned (given the fact that it extends `SerialPort.OpenOptions`).

Happy to do whatever, and I'd defer to your vote. Just wanted to explain the reasoning behind this being camel vs everything else in snake.

File: src/index.ts (hunk 4 above)
Comment (chazkiker2) [reply src/index.ts:295]: Gotcha. Well, it breaks type for `this[lock_symbol]` if ever it holds boolean. 

What is the circumstance that we'd want the `false` in there, beyond avoiding the construction of a pointless Promise? It looks like we're only ever pushing that false value (or pointless promise) if we call `this.lock()` and `this[lock_symbol]` is empty, but it's a bit unclear to me as to **why**. 

In `unlock`, we call 

```typescript
var resolve = this[lock_symbol].shift();
if (!resolve && this[lock_symbol].length) {
    resolve = this[lock_symbol].shift();
} 
resolve && resolve();
```

Which is acting as if we want to waste that first lock.

If I called
```typescript
const ryder_serial = new RyderSerial(/*port, options*/);
ryder_serial.lock(); // first time locking
// this[lock_symbol].push(false);
// return Promise.resolve();
ryder_serial.unlock(); 
// var resolve =  this[lock_symbol].shift(); // --- resolve now equals `false`, this[lock_symbol] is now empty
// resolve && resolve(); // clearly won't call, but why would we need to call it anyway?
```

I guess I just don't fully understand why `this[lock_symbol]` is holding promises at all. It seems like we use `this[lock_symbol]` as an array of arbitrary placeholders that symbolize a "lock" and simply push and pop them throughout. If it's empty? RyderSerial is unlocked! If it's not, RyderSerial is locked. 

Is there a reason that we're not simply using a `this.lock_count`?

File: src/index.ts (hunk 5 above)
Comment (chazkiker2) [reply src/index.ts:340]: Oh, this is an interesting thought...

Agreed that this can wait for later. Especially because I think the way we treat `this[train_symbol]` currently is potentially the weakest part of this code right now... I'm still motivated to refactor it into its own class but think that should wait for a different PR (this one is gettin' big).

File: test/index.js (hunk 6 above)
Comment (chazkiker2) [reply test/index.js:3]: Agreed completely. Honestly only included this to showcase that it IS working without requiring you to mess around with any `package.lock` files in the `ryder-cli-proto` to confirm haha.

I will delete this folder.

File: package.json (hunk 7 above)
Comment (chazkiker2) [reply package.json:5]: deal.

File: src/index.ts (hunk 4 above)
Comment (MarvinJanssen) [reply src/index.ts:295]: So developers can do this:

```JS
async () =>
 {
 await serial.lock(); // wait until we get a lock
 // do async stuff, other places that requested
 // a lock later will wait until we unlock.
 serial.unlock();
 }
 ```

If there are no pending lock then it resolves right away but we would still like to know a lock was requested, hence the `false`. I just added a falsy value. We can put in a `Promise.resolve` too and simplify `unlock()`.

Check out `RyderSerial.prototype.sequence()`.

File: src/index.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (chazkiker2) [reply src/index.ts:48]: TL;DR —— are we sure changing these properties to snake_case won't be a breaking change as well? These option properties were already included in the code & documented in examples.

File: src/index.ts (hunk 4 above)
Comment (chazkiker2) [reply src/index.ts:295]: Understood. That makes perfect sense. Thanks for clarifying.

Can we keep the pointless promise for now for sake of type simplification? The boolean was a beast to manage with TS.

If that's fine I'll go ahead and update `unlock` as well.

File: package.json (hunk 7 above)
Comment (chazkiker2) [reply package.json:5]: Question — how does this work for npm installation? 

If I `npm i ryderserial-proto` (though I know currently we point to github repo/branch itself) and `dist/index.js` is untracked despite being the main entry point... how is my client able to access this folder?

I'm new to distributing npm libraries so sorry if this is a stupid question.

**Comment by chazkiker2** [comment]
# Iteration 2

## Changes to Implement (will update as status progresses)

- [x] change `outDir` to `dist` and untrack.
- [x] get rid of horrible test file
- [ ] include typedefs in build
- [ ] resolve `false` vs empty Promise in `lock()` and `unlock()` methods


## TBD / Awaiting input

- [ ] resolve snake case issue for `RyderSerialOptions` (awaiting input)
  - need to know whether changing these options is a breaking change or not. If so, should we make that refactor and version bump?
- [ ] potential refactor for `TrainEntry[4]` output buffer; what's the best type option: `string`, `Array<Byte>`, `Buffer`?
  - All this is waiting on for now is my own decision, I need to dig deeper before I can speak to which we should use

## Leaving for later PRs

- linter & format on commit
- potentially queue commands in case the device disconnects; run queued command(s) once device reconnects
- refactor `this[train_symbol]` into its own class; add more formalized support & better sequencer

**Comment by chazkiker2** [comment]
@MarvinJanssen non-breaking changes that don't require a version bump are implemented. 

**TL;DR** — I vote we merge this PR in right now (unless you feel differently or notice specific bugs).

--- 

Here's my vote for right now. The PR as it stands is fully functional, bundling types into the npm package, and will not mess with any client code whatsoever. 

- I vote we leave the camel-case typing in `RyderSerialOptions` due to the fact that our documentation recommends to use it (i.e., `reconnectTime` is listed in camel-case in the very first example in `README.md`). We should definitely make all casing consistent eventually, but maybe leave it for a future PR, specifically the one that adds in `.eslint` support and maybe `.prettierrc` config. 
- I vote we also leave `TrainEntry[4]: string` due to the fact that it would require a version bump to change. I'm still unsure which is the most efficient choice, but once again this PR has gotten fairly large and I think that type refactor could be a really nice, granular PR. 

Let me know your thoughts. Happy to make those changes part of this PR if you feel strongly about it.

File: src/index.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (chazkiker2) [reply src/index.ts]: went ahead and snake-cased them anyway. May as well include consistency in this now ginormous PR. If it breaks, we bump, yeah?

File: src/index.ts [hunk 9]
```diff
@@ -0,0 +1,384 @@
...
+  debug: boolean;
+}
+
+type TrainEntry = [
+  string,                 // data
+  (value?: any) => void,  // resolve
+  (error?: any) => void,  // reject
+  boolean,                // isEscapedByte
+  // TODO: does this need to be refactored to type `Buffer`?
+  string                  // output buffer
```
[the remark this answers is not in the document]
Comment (chazkiker2) [reply src/index.ts]: At this point, I have so many comments all over the place in this PR convo. 

I mentioned this elsewhere, but for the sake of anyone who may scroll through these comments I'll reiterate here.

I think your thoughts here are very much worth considering. But this change should definitely be a part any upcoming refactor we dive into for the train/sequencer logic. 

I agree that we should hit most efficient. I don't know if `Buffer` or `Array<Buffer>` would be better, but I'm pretty confident that string is not the efficient move here. (You can see me call `Buffer.from(data[0]).toString("hex");` (where `data[0]` is of type `string` to satisfy type safety for string support but still only log HEX code). So converting from string to buffer to hex string. 

Happy to dive deeper into this problem, but I __really__ think this of all changes should be part of a separate PR that addresses various train-related issues.

File: src/ryder-serial.ts [hunk 10]
```diff
@@ -0,0 +1,493 @@
...
+  reconnect_time?: number,
+  debug?: boolean;
+}
+
+type TrainEntry = [
+  string,                 // TrainEntry[0] -- data
+  (value?: any) => void,  // TrainEntry[1] -- resolve function
+  (error?: any) => void,  // TrainEntry[2] -- reject function
+  boolean,                // TrainEntry[3] -- isEscapedByte
+  // TODO: does TrainEntry[4] (output buffer) need to be refactored to type `Buffer`, `Array<Byte>`, or stay as `string`?
```
Comment (chazkiker2) [inline src/ryder-serial.ts:60]: Leave for future PR for reasons mentioned in PR conversation.

**Comment by chazkiker2** [comment]
@MarvinJanssen this PR is ready for your review once again. It's starting to look pretty great for a foundational TS layer 👀 . 

## TL;DR

I think at this point, we should focus on getting this in a merge-able state (if it's not already there) and worry about future features in later PRs.

I think this code is pretty damn clean for a TypeScript base-layer. It certainly has parts to be improved (we're still living up to that `-proto` in the name) but I think my vision for this initial refactor has definitely been met.

## Details

I've responded to all your comments, but to save you some comment-sifting, here's the quick overview.

All implemented changes & thoughts AFTER your initial review can be found [in this comment](https://github.com/Light-Labs/ryderserial-proto/pull/5#issuecomment-846293323)

The two main issues I went back and forth on are illustrated in the bullet-point list below.

- (`TrainEntry[4]`) output buffer type -- `string` vs `Buffer` vs `Array<Buffer>`. 
  - all points you made in previous comment are great. 
  - Agreed that we should aim for fastest / most efficient. My bet is that'll be the `Buffer`. I'm fairly confident that the `string` would be the least efficient of the three. _(Though of course it's always implementation that counts)_
  - that said, (as you mentioned) this will present a breaking change (requiring a version bump) and I fear that trying to tackle this issue right now may be optimizing too early.
  - I personally vote we tackle this decision and its implementation in a separate PR that aims to overhaul the `train` / sequencer logic in several aspects.
- `Option` interface camel-case vs snake-case
  - I switched it to snake-case for consistency.
  - I am afraid this _MAY_ just be a breaking change
  - I think it definitely justifies a `README.md` update (though there are probably several parts of this PR that justify a `README` update)
  - Either way, this PR got so huge that I would _hate_ to merge it in with inconsistent casing. This PR may require a version update (idk) but either way we may as well walk out of it with consistent casing across the board.

**Review by MarvinJanssen — APPROVED** [review]
This is amazing! I agree we should get this merged and then see where we go from there.

> * `Option` interface camel-case vs snake-case
>   
>   * I switched it to snake-case for consistency.
>   * I am afraid this _MAY_ just be a breaking change
>   * I think it definitely justifies a `README.md` update (though there are probably several parts of this PR that justify a `README` update)
>   * Either way, this PR got so huge that I would _hate_ to merge it in with inconsistent casing. This PR may require a version update (idk) but either way we may as well walk out of it with consistent casing across the board.

I think the best thing we can do is to do a patch version bump (`"0.0.1"` -> `"0.0.2"`) and allow both snake casing and camel casing for the Options. We can then trigger a deprecation warning when a camel cased option is set. Thoughts?

If you don't mind to make that final change I am happy to merge this today. 👍

**Comment by chazkiker2** [comment]
@MarvinJanssen Done :)

**Comment by MarvinJanssen** [comment]
🎉
